packages/framework/tests/Feature/PaginatorTest.php

```diff
@@ -0,0 +1,296 @@
+
+        $this->expectException(InvalidArgumentException::class);
+        $service->setCurrentPage(6);
+    }
+
+    public function testCannotSetInvalidCurrentPageNumberInConstructor()
+    {
+        $this->expectException(InvalidArgumentException::class);
+        new Paginator(
+            range(1, 50),
+            (10),
+            currentPageNumber: 6
+        );
+    }
+
+    public function testLastPageReturnsTheLastPageNumber()
+    {
+        $this->assertSame(5, $this->makePaginator()->lastPage());
+    }
+
+    public function testTotalPagesReturnsTheTotalNumberOfPages()
+    {
+        $this->assertSame(5, $this->makePaginator()->totalPages());
+    }
+
+    public function testPerPageReturnsTheNumberOfItemsToBeShownPerPage()
+    {
+        $this->assertSame(10, $this->makePaginator()->perPage());
+    }
+
+    public function testHasMultiplePagesReturnsTrueIfThereAreEnoughItemsToSplitIntoMultiplePages()
+    {
+        $this->assertTrue($this->makePaginator()->hasMultiplePages());
+    }
+
+    public function testHasPagesReturnsFalseIfThereAreNotEnoughItemsToSplitIntoMultiplePages()
+    {
+        $this->assertFalse($this->makePaginator(1, 9)->canNavigateForward());
+    }
+
+    public function testHasMorePagesReturnsTrueIfCursorCanNavigateForward()
+    {
+        $this->assertTrue($this->makePaginator()->canNavigateForward());
+    }
+
+    public function testHasMorePagesReturnsFalseIfCursorCannotNavigateForward()
+    {
+        $this->assertFalse($this->makePaginator()->setCurrentPage(5)->canNavigateForward());
+    }
+
+    public function testHasFewerPagesReturnsTrueIfCursorCanNavigateBack()
+    {
+        $this->assertTrue($this->makePaginator()->setCurrentPage(2)->canNavigateBack());
+    }
+
+    public function testHasFewerPagesReturnsFalseIfCursorCannotNavigateBack()
+    {
+        $this->assertFalse($this->makePaginator()->canNavigateBack());
+    }
+
+    public function testPreviousMethodWithoutFewerPagesReturnsFalse()
+    {
+        $this->assertFalse($this->makePaginator()->previous());
+    }
+
+    public function testNextMethodWithoutMorePagesReturnsFalse()
+    {
+        $this->assertFalse($this->makePaginator()->setCurrentPage(5)->next());
+    }
+
+    public function testPreviousMethodReturnsPreviousPageLinkWhenNoBaseRouteIsSet()
+    {
+        $this->assertSame('page-1.html', $this->makePaginator()->setCurrentPage(2)->previous());
+    }
+
+    public function testNextMethodReturnsNextPageLinkWhenNoBaseRouteIsSet()
+    {
+        $this->assertSame('page-2.html', $this->makePaginator()->setCurrentPage(1)->next());
+    }
+
+    public function testPreviousAndNextMethodsWithBaseRouteSet()
+    {
+        $pages[1] = new VirtualPage('pages/page-1');
+        $pages[2] = new VirtualPage('pages/page-2');
+        $pages[3] = new VirtualPage('pages/page-3');
+        $pages[4] = new VirtualPage('pages/page-4');
+        $pages[5] = new VirtualPage('pages/page-5');
```

Comment:
I wonder if we could add a helper method that generates the paginated virtual pages / blade pages (using a specified template) this could then be used in a provider when booting to inject those pages into the page collection.